Source/WebKit/Shared/Extensions/WebExtensionLocalization.cpp

```diff
@@ -0,0 +1,349 @@
+
+    localizedString = stringByReplacingNamedPlaceholdersInString(localizedString, namedPlaceholders);
+    localizedString = stringByReplacingPositionalPlaceholdersInString(localizedString, placeholders);
+
+    localizedString = localizedString.impl()->replace("$$"_s, "$"_s);
+
+    return localizedString;
+}
+
+void WebExtensionLocalization::localizedArrayForArray(RefPtr<JSON::Array> json)
+{
+    if (!json)
+        return;
+
+    for (auto value : *json) {
+        auto valueType = value->type();
+
+        if (valueType == JSON::Value::Type::String)
+            value->asString() = localizedStringForString(value->asString());
+        else if (valueType == JSON::Value::Type::Array)
+            localizedArrayForArray(value->asArray());
+        else if (valueType == JSON::Value::Type::Object)
+            localizedJSONforJSON(value->asObject());
+    }
+}
+
+String WebExtensionLocalization::localizedStringForString(String sourceString)
+{
+    String localizedString = sourceString;
+
+    JSC::Yarr::RegularExpression localizableStringRegularExpression = JSC::Yarr::RegularExpression("__MSG_([A-Za-z0-9_@]+)__"_s, { });
+
+    int index = 0;
+    while (index < static_cast<int>(localizedString.length())) {
+        int matchLength;
+        index = localizableStringRegularExpression.match(localizedString, index, &matchLength);
+        if (index < 0)
+            break;
+
+        String key = localizedString.substring(index, matchLength);
+        String localizedReplacement = localizedStringForKey(key);
+
+        localizedString = makeStringByReplacingAll(localizedString, key, localizedReplacement);
+
+        index += localizedReplacement.length();
+    }
+
+    return localizedString;
+}
+
+RefPtr<JSON::Object> WebExtensionLocalization::localizationJSONForWebExtension(WebExtension& webExtension, const String& withLocale)
+{
+    StringBuilder pathBuilder;
+    pathBuilder.append("_locales/"_s);
+    pathBuilder.append(withLocale);
+    pathBuilder.append("/messages.json"_s);
+    String path = pathBuilder.toString();
+
+    RefPtr<API::Error> error;
+    RefPtr data = webExtension.resourceDataForPath(path, error, WebExtension::CacheResult::No, WebExtension::SuppressNotFoundErrors::Yes);
+    if (!data || error) {
+        webExtension.recordErrorIfNeeded(error);
+        return nullptr;
+    }
+
+    return JSON::Value::parseJSON(String::fromUTF8(data->span()))->asObject();
+}
+
+Ref<JSON::Object> WebExtensionLocalization::predefinedMessages(RefPtr<JSON::Object> localizedJSON)
+{
+    Ref<JSON::Object> predefinedMessages = JSON::Object::create();
+
+    auto createMessageKey = [](String messageValue) {
+        Ref<JSON::Object> object = JSON::Object::create();
+        object->setString(messageKey, messageValue);
+        return object;
+    };
+
+    if (localizedJSON && !m_localeString.isEmpty()) {
+        if (auto firstMessageObject = localizedJSON->getObject(*(localizedJSON->keys().begin()))) {
```

Comment:
```suggestion
        if (RefPtr firstMessageObject = localizedJSON->getObject(*(localizedJSON->keys().begin()))) {
```

You should check `localizedJSON->keys()` is not empty, as this would fail if it was.